File: src/main/kotlin/by/mksn/kwitapi/configuration/security/JdbcUserDetailsService.java
package by.mksn.kwitapi.configuration.security;

import by.mksn.kwitapi.entity.User;
import by.mksn.kwitapi.service.UserService;
import org.springframework.security.core.authority.AuthorityUtils;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;

public class JdbcUserDetailsService implements UserDetailsService {

    private final UserService userService;

    public JdbcUserDetailsService(UserService userService) {
        this.userService = userService;
    }

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        try {
            username = URLDecoder.decode(username, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        User user = userService.findByUsername(username);
        if (user == null) {
            throw new UsernameNotFoundException("User " + username + " not found in database");
        }
        return new org.springframework.security.core.userdetails.User(
                user.getUsername(),
                user.getPasswordHash(),
                !user.isDeleted(),
                true,
                true,
                true,
                AuthorityUtils.createAuthorityList("USER", "write")
        );
    }
}

File: src/main/kotlin/by/mksn/kwitapi/configuration/security/JdbcUserDetailsService.kt
package by.mksn.kwitapi.configuration.security

import by.mksn.kwitapi.DEFAULT_ENCODING
import by.mksn.kwitapi.repository.UserRepository
import org.slf4j.LoggerFactory
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import java.net.URLDecoder

class JdbcUserDetailsService(
        val userRepository: UserRepository
) : UserDetailsService {

    companion object {
        val logger = LoggerFactory.getLogger(JdbcUserDetailsService::class.java)!!
    }

    override fun loadUserByUsername(email: String): org.springframework.security.core.userdetails.UserDetails {
        val decodedEmail = URLDecoder.decode(email, DEFAULT_ENCODING)
        val user = userRepository.findByEmail(decodedEmail)
        if (user == null || user.isDeleted) {
            logger.info("User with email $decodedEmail not found")
            throw UsernameNotFoundException("User with email $decodedEmail not found")
        }
        return UserDetails(user)
    }
}

